Authentication, authorization and edge configuration through a LDAP server in the backend.

chpoulter opened the pull request:
After building some promising docker images for running OpenEMS in my local environment I missed a connection to my LDAP. So I just wrote this code down for authentication, authorization and edge configuration through a LDAP server for the backend part of OpenEMS and it has been running smoothly the last few days. I probably will also add support for the edge soon. Hope it might be of interest.

miettal commented on io.openems.backend.metadata.ldap/bnd.bnd:
```diff
@@ -0,0 +1,16 @@
+Bundle-Vendor: Christian Poulter
+Bundle-ContactAddress: devel(at)poulter.de
+Bundle-Copyright: OpenEMS Backend Metadata LDAP (C) 2024 Christian Poulter
+Bundle-License: https://www.gnu.org/licenses/agpl-3.0.txt
```
Mixing AGPL License and EPL License codes in one binary is very difficult to use.
Because AGPL infect to other files, and it require redistribution these files by AGPL License.
This means existing EPL license code need to change AGPL license.
OpenEMS is OSS software and authors are many. So it is difficult to get permission to change license from ALL original content right holders(authors).

I think better to change license to EPL, if you can agree it.

miettal replied:
// I feel this PR is very very nice! and exciting. and I don't hate AGPL/EPL License.
// I love OSS license and culture, so I just pointed out the license problem.

chpoulter replied:
Hi @miettal ... thanks for your feedback. I am not so familiar with all these OSS licenses. At the moment reading EPL in opposite to GPL feels a bit like "you can contribute for free, but my modules stay closed". I will think about it and might change the PR.

miettal replied:
I found useful EPL official FAQ. I share it. This kind matter is complex and not easy to understand... I'm also studying now.

https://www.eclipse.org/legal/epl-2.0/faq.php#h.5ucozq4kvv7o

> [4.14. Does the EPL allow me to take the Source Code for a Program licensed under it and include all or part of it in another program licensed under the GNU General Public License (GPL), Berkeley Software Distribution (BSD) license or other Open Source license?](https://www.eclipse.org/legal/epl-2.0/faq.php#h.5ucozq4kvv7o)
No. Only the owner of software can decide whether and how to license it to others. Contributors to a Program licensed under the EPL understand that source code for the Program will be made available under the terms of the EPL. Unless you are the owner of the software or have received permission from the owner, you are not authorized to apply the terms of another license to the Program by including it in a program licensed under another Open Source license.

> However, if the content has been made available under the EPL-2.0 with Secondary License, the content may be distributed under the terms of that Secondary License while maintaining the original code base under the EPL-2.0.

> [4.15. Is my EPL-2.0 licensed project GPL compatible by default?](https://www.eclipse.org/legal/epl-2.0/faq.php#h.fn86eyf5mvhs)
No. EPL-2.0 licensed projects may explicitly add GPL compatibility by way of adding a Secondary License as per Exhibit A — Form of Secondary Licenses Notice. See Section 3.2 of the EPL-2.0. Please note adding a secondary license is considered a relicensing exercise and all copyright holders must agree to the license change.

note: This PR and code is your content. Deciding a license for this code is your right. Anyone cannot force it.

chpoulter commented:
I changed the license to EPL to match the other code and fixed @dhbfischer's suggestions. Please note that the memberOf in my LDAP is auto-generated by an overlay. I added a note on this to the readme.

chpoulter commented:
adopted to modified interface Metadata

chpoulter commented:
added LDAP support in edge

chpoulter commented on io.openems.backend.metadata.ldap/src/io/openems/backend/metadata/ldap/LdapUserReader.java:
```diff
@@ -0,0 +1,260 @@
+            }
+
+            // globalRole
+            final List<String> memberOf = LdapUtils.extractValuesFromAttributes(resultAttributes, "memberOf", username);
```
memberOf is generated and maintained automatically if you install the memberOf overlay. Otherwise you have to create it on your own. I've added a note in the README on this.

wdr-daan replied:
Thanks, I managed to this and then it works!

dhbfischer replied:
Thanks, I've managed to do this and it works!

chpoulter commented:
adopted to current develop branch and improved login page

chpoulter commented:
but looks like this ldap contribution is not really of interest ...

chpoulter commented:
I've stopped managing this branch as I moved code to a maven project. If somebody is interested, you can find it in my [OpenEMS Addons](https://github.com/chpoulter/openems-addons/).
